pkg/metricsadapter/provider/custommetrics.go

```diff
@@ -3,33 +3,353 @@ package provider
+		return nil, err
+	}
+	return metricsConvertCustomMetricsV1beta2ToInternalCustomMetrics(metricObj)
+}
+
+func (c *CustomMetricsProvider) getForNamespace(ctx context.Context, clusterName, namespace string, metricName string, metricSelector labels.Selector) (*custom_metrics.MetricValueList, error) {
+	discoveryClient := c.multiClusterDiscovery.Get(clusterName)
+	if discoveryClient == nil {
+		err := fmt.Errorf("failed to get MultiClusterDiscovery for cluster(%s)", clusterName)
+		klog.Error(err)
+		return nil, err
+	}
+	version, err := c.getPreferredVersion(discoveryClient)
+	if err != nil {
+		klog.Errorf("failed to get custom.metrics.k8s.io preferred version for cluster(%s),Error: %v", clusterName, err)
+		return nil, err
+	}
+	params, err := versionConverter.ConvertListOptionsToVersion(&custom_metrics.MetricListOptions{
+		MetricLabelSelector: metricSelector.String(),
+	}, version)
+	if err != nil {
+		return nil, err
+	}
+	result := discoveryClient.RESTClient().Get().Prefix("/apis/"+version.String()).
+		Resource("metrics").
+		Namespace(namespace).
+		Name(metricName).
+		SpecificallyVersionedParams(params, custommetricsschema.ParameterCodec, version).
+		Do(ctx)
+	metricObj, err := versionConverter.ConvertResultToVersion(result, custommetricsv1beta2.SchemeGroupVersion)
+	if err != nil {
+		return nil, err
+	}
+	return metricsConvertCustomMetricsV1beta2ToInternalCustomMetrics(metricObj)
 }
 
 // ListAllMetrics returns all metrics in all member clusters
 func (c *CustomMetricsProvider) ListAllMetrics() []provider.CustomMetricInfo {
-	return []provider.CustomMetricInfo{}
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return []provider.CustomMetricInfo{}
+	}
+	var customMetricInfos []provider.CustomMetricInfo
+	metricInfoChan := make(chan provider.CustomMetricInfo)
+	wg := sync.WaitGroup{}
+	for _, cluster := range clusters {
+		wg.Add(1)
+		go func(clusterName string) {
+			defer wg.Done()
+			discoveryClient := c.multiClusterDiscovery.Get(clusterName)
+			if discoveryClient == nil {
+				err := fmt.Errorf("failed to get MultiClusterDiscovery for cluster(%s)", clusterName)
+				klog.Error(err)
+				return
+			}
+			apiGroups, err := discoveryClient.ServerGroups()
+			if err != nil {
+				klog.Errorf("Failed to query resource in cluster(%s): %+v", clusterName, err)
+				return
+			}
+			supportGroupVersion, support := supportedMetricsAPIVersionAvailable(apiGroups)
+			if !support {
+				klog.Warningf("custom.metrics.k8s.io not found in cluster(%s)", clusterName)
+				return
+			}
+			resources, err := discoveryClient.ServerResourcesForGroupVersion(supportGroupVersion.String())
+			if err != nil {
+				klog.Warningf("Failed to query %s resource in cluster(%s): %+v", supportGroupVersion.String(), clusterName, err)
+				return
+			}
+			for _, resource := range resources.APIResources {
+				groupResourceAndMetricName := strings.SplitN(resource.Name, "/", 2)
```

Comment:
Can we add some comments how the `resource.Name` looks like?

Reply:
Added.